visual-js/visual/src/utils.ts

```diff
@@ -20,3 +28,92 @@ export const isSkipMode = (): boolean => {
   const VISUAL_SKIP_ENV_VAR = 'SAUCE_VISUAL_SKIP';
   return Boolean(process.env[VISUAL_SKIP_ENV_VAR]);
 };
+
+export const makeValidate =
+  <T extends Type<any>>(t: T) =>
+  (x: unknown): T['infer'] => {
+    const { data, problems } = t(x);
+    if (problems) {
+      throw new Error(problems.toLocaleString());
+    }
+    return data;
+  };
+
+const ignoreRegionType: Type<RegionIn> = type({
+  width: 'integer',
+  height: 'integer',
+  x: 'integer',
+  y: 'integer',
+  'name?': 'string | null | undefined',
+});
+
+export const isIgnoreRegion = ignoreRegionType.allows;
+
+export const validateIgnoreRegion = makeValidate(ignoreRegionType);
+
+export const getDiffingOptions = <T>(
+  region: RegionType<T>,
+): DiffingOptionsIn | undefined => {
+  const { element: __, ...rest } = region;
+  if ('enableOnly' in rest || 'disableOnly' in rest) {
+    return selectiveRegionOptionsToDiffingOptions(rest);
+  }
+  return undefined;
+};
+
+/**
+ * Parse ignore elements, regions, and selective regions and format them for use with the API
+ * @param ignore
+ * @param isElement A callback to check whether an item is a web element, depending on framework implementation.
+ * @param getElementMeta A callback to gather the required metadata from an element for use with the API.
+ */
+export const parseRegionsForAPI = async <T>(
+  ignore: RegionType<T>[],
+  isElement: (element: unknown) => element is T,
+  getElementMeta: (
+    element: T,
+  ) => { id: string; name?: string } | Promise<{ id: string; name?: string }>,
+): Promise<{
+  ignoreRegions: RegionIn[];
+  ignoreElements: ElementIn[];
+}> => {
+  const ignoreElements: ElementIn[] = [];
+  const ignoreRegions: RegionIn[] = [];
+
+  const pushElement = async (region: RegionType<T>) => {
+    const { element: maybePromiseElement, ...other } = region;
+    const element = await maybePromiseElement;
+
+    if (Array.isArray(element)) {
+      for await (const subElement of element) {
+        await pushElement({
+          element: subElement,
+          ...other,
+        });
+      }
```

Comment:
There is still the recursion that can be eliminated (to make it easier for future editors and reviewers), because we only have two nested layers.

I also think, we only need to pass one function that checks and resolves at the same time:

```ts
resolveElement: (maybeElement: unknown) => Promise<{id: string, name?: string}[] | null>
```